lib/modules/manager/npm/artifacts.ts

```diff
@@ -0,0 +1,77 @@
+import { logger } from '../../../logger';
+import { exec } from '../../../util/exec';
+import { ExecOptions } from '../../../util/exec/types';
+import { readLocalFile } from '../../../util/fs';
+import { UpdateArtifact, UpdateArtifactsResult } from '../types';
+
+// used to add the digest recommened by corepack
+// since the digest is generated and added by corepack itself and cannot be fecthed from any npm regsitry
+export async function updateArtifacts({
+  packageFileName,
+  config,
+  updatedDeps,
+  newPackageFileContent: existingPackageFileContent,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`npm.updateArtifacts(${packageFileName})`);
+  const packageManagerUpdate = updatedDeps.find(
+    (dep) => dep.depType === 'packageManager',
+  );
+
+  if (!packageManagerUpdate) {
```

Comment:
also don't run it if no hash is present. run it on pin deps if no hash is present